File: sdk/typescript/src/contracts/dispute.ts
import { BaseContract } from './base';
import {
  DisputeParams,
  ResolutionOutcome,
  Juror,
  Vote,
  Evidence,
  Appeal,
  EthereumAddress,
  Bytes32,
  CallOverrides,
  TransactionResult,
} from '../types';
import { EthersProvider } from '../providers';
import {
  DisputeDAOABI,
  DisputeResolutionABI,
  DisputeJuryABI,
  DisputeVotingABI,
  DisputeEvidenceABI,
  DisputeAppealABI,
} from '../abis';

export class DisputeDAO extends BaseContract {
  constructor(address: EthereumAddress, provider: EthersProvider) {
    super(address, DisputeDAOABI, provider);
  }

  async updateParams(
    params: DisputeParams,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('updateParams', [params], overrides);
  }

  async getParams(): Promise<DisputeParams> {
    return this.call('getParams', []);
  }

  async withdrawTreasury(
    token: EthereumAddress,
    recipient: EthereumAddress,
    amount: bigint,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('withdrawTreasury', [token, recipient, amount], overrides);
  }
}

export class DisputeResolution extends BaseContract {
  constructor(address: EthereumAddress, provider: EthersProvider) {
    super(address, DisputeResolutionABI, provider);
  }

  async resolveDispute(
    disputeId: bigint,
    outcome: ResolutionOutcome,
    detailsHash: Bytes32,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send(
      'resolveDispute',
      [disputeId, outcome, detailsHash],
      overrides
    );
  }

  async executeResolution(
    disputeId: bigint,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('executeResolution', [disputeId], overrides);
  }

  async getResolution(
    disputeId: bigint
  ): Promise<{ outcome: ResolutionOutcome; detailsHash: Bytes32; executed: boolean }> {
    return this.call('getResolution', [disputeId]);
  }

  async canAppeal(disputeId: bigint): Promise<boolean> {
    return this.call('canAppeal', [disputeId]);
  }
}

export class DisputeJury extends BaseContract {
  constructor(address: EthereumAddress, provider: EthersProvider) {
    super(address, DisputeJuryABI, provider);
  }

  async registerJuror(
    stake: bigint,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('registerJuror', [stake], overrides);
  }

  async unregisterJuror(overrides?: CallOverrides): Promise<TransactionResult> {
    return this.send('unregisterJuror', [], overrides);
  }

  async selectJury(
    disputeId: bigint,
    jurySize: bigint,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('selectJury', [disputeId, jurySize], overrides);
  }

  async slashJuror(
    juror: EthereumAddress,
    amount: bigint,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('slashJuror', [juror, amount], overrides);
  }

  async getJuror(account: EthereumAddress): Promise<Juror> {
    const result = await this.call<any>('getJuror', [account]);
    return {
      account: result.account,
      stake: result.stake,
      selectionScore: result.selectionScore,
      active: result.active,
    };
  }

  async getJurorsForDispute(disputeId: bigint): Promise<EthereumAddress[]> {
    return this.call('getJurorsForDispute', [disputeId]);
  }
}

export class DisputeVoting extends BaseContract {
  constructor(address: EthereumAddress, provider: EthersProvider) {
    super(address, DisputeVotingABI, provider);
  }

  async castVote(
    disputeId: bigint,
    choice: number,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('castVote', [disputeId, choice], overrides);
  }

  async closeVoting(
    disputeId: bigint,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('closeVoting', [disputeId], overrides);
  }

  async getVote(disputeId: bigint, voter: EthereumAddress): Promise<Vote> {
    const result = await this.call<any>('getVote', [disputeId, voter]);
    return {
      voter: result.voter,
      choice: result.choice,
      weight: result.weight,
      timestamp: result.timestamp,
    };
  }

  async getVoteTally(disputeId: bigint): Promise<[bigint, bigint, bigint]> {
    return this.call('getVoteTally', [disputeId]);
  }

  async hasVoted(disputeId: bigint, voter: EthereumAddress): Promise<boolean> {
    return this.call('hasVoted', [disputeId, voter]);
  }

  async getVotingEndTime(disputeId: bigint): Promise<bigint> {
    return this.call('getVotingEndTime', [disputeId]);
  }
}

export class DisputeEvidence extends BaseContract {
  constructor(address: EthereumAddress, provider: EthersProvider) {
    super(address, DisputeEvidenceABI, provider);
  }

  async submitEvidence(
    disputeId: bigint,
    evidenceHash: Bytes32,
    metadataHash: Bytes32,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send(
      'submitEvidence',
      [disputeId, evidenceHash, metadataHash],
      overrides
    );
  }

  async getEvidence(evidenceId: bigint): Promise<Evidence> {
    const result = await this.call<any>('getEvidence', [evidenceId]);
    return {
      evidenceId: result.evidenceId,
      disputeId: result.disputeId,
      submitter: result.submitter,
      evidenceHash: result.evidenceHash,
      metadataHash: result.metadataHash,
      submittedAt: result.submittedAt,
    };
  }

  async getEvidenceByDispute(disputeId: bigint): Promise<bigint[]> {
    return this.call('getEvidenceByDispute', [disputeId]);
  }

  async getEvidencePeriodEnd(disputeId: bigint): Promise<bigint> {
    return this.call('getEvidencePeriodEnd', [disputeId]);
  }
}

export class DisputeAppeal extends BaseContract {
  constructor(address: EthereumAddress, provider: EthersProvider) {
    super(address, DisputeAppealABI, provider);
  }

  async fileAppeal(
    disputeId: bigint,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('fileAppeal', [disputeId], overrides);
  }

  async resolveAppeal(
    appealId: bigint,
    status: number,
    overrides?: CallOverrides
  ): Promise<TransactionResult> {
    return this.send('resolveAppeal', [appealId, status], overrides);
  }

  async getAppeal(appealId: bigint): Promise<Appeal> {
    const result = await this.call<any>('getAppeal', [appealId]);
    return {
      appealId: result.appealId,
      disputeId: result.disputeId,
      appellant: result.appellant,
      bond: result.bond,
      appealedAt: result.appealedAt,
      status: result.status,
    };
  }

  async getAppealsByDispute(disputeId: bigint): Promise<bigint[]> {
    return this.call('getAppealsByDispute', [disputeId]);
  }

  async getAppealPeriod(): Promise<bigint> {
    return this.call('getAppealPeriod', []);
  }

  async getAppealBond(): Promise<bigint> {
    return this.call('getAppealBond', []);
  }
}

File: sdk/python/covenant_sdk/contracts/dispute.py
"""Dispute contract wrappers."""

from typing import Dict, List
from web3 import Web3

from .base import BaseContract
from ..types import EthereumAddress, Bytes32, DisputeParams, ResolutionOutcome, AppealStatus
from ..abis import DISPUTE_DAO_ABI, DISPUTE_RESOLUTION_ABI, DISPUTE_APPEAL_ABI
from ..utils import TxReceipt


class DisputeDAO(BaseContract):
    """Wrapper for the DisputeDAO contract."""

    def __init__(self, address: EthereumAddress, w3: Web3):
        super().__init__(address, DISPUTE_DAO_ABI, w3)

    def update_params(self, params: DisputeParams) -> TxReceipt:
        """Update dispute parameters."""
        param_tuple = (params.min_stake, params.voting_period, params.quorum, params.appeal_threshold)
        return self._send_transaction("updateParams", param_tuple)

    def get_params(self) -> DisputeParams:
        """Get current dispute parameters."""
        result = self._call("getParams")
        return DisputeParams(
            min_stake=result[0],
            voting_period=result[1],
            quorum=result[2],
            appeal_threshold=result[3]
        )

    def withdraw_treasury(self, token: EthereumAddress, recipient: EthereumAddress, amount: int) -> TxReceipt:
        """Withdraw funds from the treasury."""
        return self._send_transaction("withdrawTreasury", token, recipient, amount)


class DisputeResolution(BaseContract):
    """Wrapper for the DisputeResolution contract."""

    def __init__(self, address: EthereumAddress, w3: Web3):
        super().__init__(address, DISPUTE_RESOLUTION_ABI, w3)

    def resolve_dispute(self, dispute_id: int, outcome: ResolutionOutcome, details_hash: Bytes32) -> TxReceipt:
        """Resolve a dispute with an outcome."""
        return self._send_transaction("resolveDispute", dispute_id, outcome.value, details_hash)

    def execute_resolution(self, dispute_id: int) -> TxReceipt:
        """Execute a resolved dispute."""
        return self._send_transaction("executeResolution", dispute_id)

    def get_resolution(self, dispute_id: int) -> Dict:
        """Get resolution details."""
        result = self._call("getResolution", dispute_id)
        return {
            "outcome": ResolutionOutcome(result[0]),
            "details_hash": result[1],
            "executed": result[2],
        }

    def can_appeal(self, dispute_id: int) -> bool:
        """Check if a dispute can be appealed."""
        return self._call("canAppeal", dispute_id)


class DisputeAppeal(BaseContract):
    """Wrapper for the DisputeAppeal contract."""

    def __init__(self, address: EthereumAddress, w3: Web3):
        super().__init__(address, DISPUTE_APPEAL_ABI, w3)

    def file_appeal(self, dispute_id: int, value: int = 0) -> TxReceipt:
        """File an appeal for a dispute."""
        return self._send_transaction("fileAppeal", dispute_id, value=value)

    def resolve_appeal(self, appeal_id: int, status: AppealStatus) -> TxReceipt:
        """Resolve an appeal."""
        return self._send_transaction("resolveAppeal", appeal_id, status.value)

    def get_appeal(self, appeal_id: int) -> Dict:
        """Get appeal details."""
        result = self._call("getAppeal", appeal_id)
        return {
            "appeal_id": result[0],
            "dispute_id": result[1],
            "appellant": result[2],
            "bond": result[3],
            "appealed_at": result[4],
            "status": AppealStatus(result[5]),
        }

    def get_appeals_by_dispute(self, dispute_id: int) -> List[int]:
        """Get all appeals for a dispute."""
        return self._call("getAppealsByDispute", dispute_id)

    def get_appeal_period(self) -> int:
        """Get the appeal period duration."""
        return self._call("getAppealPeriod")

    def get_appeal_bond(self) -> int:
        """Get the required appeal bond."""
        return self._call("getAppealBond")
